core/src/banking_trace.rs

```diff
@@ -0,0 +1,510 @@
+    enabled_tracer: Option<(
+        Sender<TimedTracedEvent>,
+        Option<JoinHandle<TracerThreadResult>>,
+        Arc<AtomicBool>,
+    )>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct TimedTracedEvent(std::time::SystemTime, TracedEvent);
+
+#[derive(Serialize, Deserialize, Debug)]
+enum TracedEvent {
+    Bank(Slot, u32, BankStatus, usize),
+    PacketBatch(ChannelLabel, BankingPacketBatch),
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+enum BankStatus {
+    Started,
+    Ended,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, Copy)]
+pub enum ChannelLabel {
+    NonVote,
+    TpuVote,
+    GossipVote,
+    Dummy,
+}
+
+struct RollingConditionGrouped {
+    basic: RollingConditionBasic,
+    is_checked: bool,
+}
+
+impl RollingConditionGrouped {
+    fn new(basic: RollingConditionBasic) -> Self {
+        Self {
+            basic,
+            is_checked: bool::default(),
+        }
+    }
+
+    fn reset(&mut self) {
+        self.is_checked = false;
+    }
+}
+
+struct GroupedWriter<'a> {
+    now: DateTime<Local>,
+    underlying: &'a mut RollingFileAppender<RollingConditionGrouped>,
+}
+
+impl<'a> GroupedWriter<'a> {
+    fn new(underlying: &'a mut RollingFileAppender<RollingConditionGrouped>) -> Self {
+        Self {
+            now: Local::now(),
+            underlying,
+        }
+    }
+}
+
+impl RollingCondition for RollingConditionGrouped {
+    fn should_rollover(&mut self, now: &DateTime<Local>, current_filesize: u64) -> bool {
+        if !self.is_checked {
+            self.is_checked = true;
+            self.basic.should_rollover(now, current_filesize)
+        } else {
+            false
+        }
+    }
+}
+
+impl<'a> Write for GroupedWriter<'a> {
+    fn write(&mut self, buf: &[u8]) -> std::result::Result<usize, io::Error> {
+        self.underlying.write_with_datetime(buf, &self.now)
+    }
+    fn flush(&mut self) -> std::result::Result<(), io::Error> {
+        self.underlying.flush()
+    }
+}
+
+pub fn sender_overhead_minimized_receiver_loop<T, E, const SLEEP_MS: u64>(
+    exit: Arc<AtomicBool>,
+    receiver: Receiver<T>,
+    mut on_recv: impl FnMut(T) -> Result<(), E>,
+) -> Result<(), E> {
+    'outer: while !exit.load(Ordering::Relaxed) {
+        'inner: loop {
+            // avoid futex-based blocking here, otherwise a sender would have to
+            // wake me up at a syscall cost...
+            match receiver.try_recv() {
+                Ok(message) => on_recv(message)?,
+                Err(TryRecvError::Empty) => break 'inner,
+                Err(TryRecvError::Disconnected) => {
+                    break 'outer;
+                }
+            };
+            if exit.load(Ordering::Relaxed) {
+                break 'outer;
+            }
+        }
+        sleep(Duration::from_millis(SLEEP_MS));
+    }
+
+    Ok(())
+}
+
+impl BankingTracer {
+    pub fn new(maybe_config: Option<(PathBuf, Arc<AtomicBool>, u64)>) -> Result<Self, TraceError> {
```

Comment:
Would be nice to have a doc comment here explaining what the pieces of config are. Particularly it's not obvious what the u64 is supposed to be unless I read the internals.